### DISCUSSION TITLE
Episodes localization and translation system

### DISCUSSION THREAD

**Discussion by Wohlstand** [Wohlstand/TheXTech#340] [Planning]
For a long time, I keep lots of ideas to make an ability to translate episodes into different languages and keep their content unchanged (I.e. make the translation system to be fully non-destructive), and I had many thinks about how I could realize this idea into reality. So, I started to think, about which format can be used for the episodes translation system. I do think, let's try to use the [JSON-based](https://docs.weblate.org/en/latest/formats.html#json-files) one.

The thing will require to build the next helper tools at the Moondust Project side:
- **Updater**, the utility that will scan the content of an episode (levels, world maps, scripts) and will generate JSON files and fill them with extracted text phrases (the file related to the original language that will contain a copy of all original phrases; and translation files that will get blank fields within copies of original phrases for human editing convenience). Updater is needed when an episode took any content changes and there is needed to add new-added fields and get rid of strings that were removed from the episode, etc.
- **Translation tool**: this is the GUI tool for translators where they will work. It will build the tree by levels/worlds, and by the source of the string. For convenience, chains of messages (spawned by events that trigger each other) will be connected to keep in-game dialogs normally visible.

What about the format of the file, I want to show the next drafts, made on a quick hand:

Example of original language file:
```json
{
    "level 1.lvlx" : 
    {
        "npc-233-talk" : "I am a warrior!",
        "npc-542-talk" : "Really?",
        "event-42-msg" : "I am strong!",
        "event-11-msg" : "I bet, I'll defeat you!",
        "event-12-msg" : "Do you even see me?",
        "script-tutorial1" : "Come to warrior and talk with him!",
        "chains" : [
            ["npc-233-talk", "event-42-msg"],
            ["npc-542-talk", "event-11-msg", "event-12-msg"]
        ]
    },
    "level 2.lvlx" : 
    {},
    "level 3.lvlx" : 
    {},
    "world.lvlx" : 
    {
        "level-12-tit" : "Battlefield",
        "level-14-tit" : "Castle"
    }
}
```

An example of a possible translation file:
```json
{
    "level 1.lvlx" : 
    {
        "npc-233-talk" : "Я воин!",
        "npc-233-talk-orig" : "I am a warrior!",
        "npc-542-talk" : "Серьёзно?",
        "npc-542-talk-orig" : "Really?",
        "event-42-msg" : "Я сильный!",
        "event-42-msg-orig" : "I am strong!",
        "event-11-msg" : "Я спорю, что одлею тебя!",
        "event-11-msg-orig" : "I bet, I'll defeat you!",
        "event-12-msg" : "Ты меня вообще видишь?",
        "event-12-msg-orig" : "Do you even see me?",
        "script-tutorial1" : "Подойди к воину и поговори с ним!",
        "script-tutorial1-orig" : "Come to warrior and talk with him!",
        "chains" : [
            ["npc-233-talk", "event-42-msg"],
            ["npc-542-talk", "event-11-msg", "event-12-msg"]
        ],
        "glossary" : [
            {"This is skeleton!" : "Это скелет!"}
        ]
    },
    "level 2.lvlx" : 
    {},
    "level 3.lvlx" : 
    {},
    "world.lvlx" : 
    {
        "level-12-tit" : "Поле боя",
        "level-12-tit-orig" : "Battlefield",
        "level-14-tit" : "Замок",
        "level-14-tit-orig" : "Castle"
    }
}
```

- Root object will contain file names
- At each file, the list of embedded strings is assigned to a specific object, such as an NPC talk message, event message, etc. In scripts is possible to use a special command format to translate the string, such as `tr("key", "The phrase")`. From Autocode scripts, source strings, block, and line numbers will be taken to identify the string.
- At the translation file, there are additional entries with `-orig` suffix to hold the copy of the original string for the convenience of file editing by humans.
- The "chains" section will contain a list of keys that triggers each other, making a sort of dialog. They can be used in the translation tool to show the line of dialog to help the translator understand the order of messages.
- The "glossary" will take the list of "source=translation" left after the deletion of strings from the episode after editing, they can be used by the tool to show the list of possible translation hints when similar cases occur.

You can express/suggest anything if you want. This is basically a planning discussion, the actual localization system will be started to be constructed after the font engine will be updated.

**Comment by 0lhi** [comment]
Looks good. Only thing I have to say is that the language files should be stored inside ```/episodefolder/languages/german.json```, since the episode root is already filled enough as it is.

**Reply by Wohlstand** [reply]
That right

**Comment by LoveBodhi** [comment]
I just made two different languages of dialogues via event for reserved purpose...

**Comment by LoveBodhi** [comment]
But, how about switchable languages?

**Comment by ds-sloth** [comment]
To continue the discussion here, I wonder if there is some way to avoid the dependence on original array indexes. Suppose the episode creator makes an update to the episode. It would be nice for existing translation files to be automatically updatable.

Maybe this can be done using the `-orig` suffixes Wohlstand was suggesting.

**Reply by Wohlstand** [reply]
Also, I'll need to add the mark that translation needs an update (as Qt Linguist does: when source string got changed, then, affected translations gets the "needs an attention" mark)

**Comment by LoveBodhi** [comment]
IDK will this work on single level?

**Reply by Wohlstand** [reply]
For single level works too in a condition that `translation_*.json` file gets placed at the same directory as level file. So, the single causal level should be placed at the isolated directory. I could try to add an ability to load the translation from the sub-directory too.

**Comment by Wohlstand** [comment]
Okay, there are othere examples based on the actual format of episode translations now:
(the detailed specification I'll write on the Wiki).

## Cache of origin (Used by translation tool itself to preserve a lot of meta-data):
Important note: The "episode title" and "episode world" field must begin with a space: it's a workaround for QMap to force-bring these fields to the top of the generated file, making easier SAX-parsing of them during the episodes list building.
```json
{
    " episode_title": "The Princess Cliche",
    " episode_world": "world.wldx",
    "bowsers_castle.lvlx": {
        "dialogues": [
            {
                "messages": [
                    "npc-2-talk"
                ],
                "note": "Name of the dialogue, a developer's hint for the translators"
            },
            {
                "messages": [
                    "event-5-msg",
                    "event-12-msg"
                ],
                "note": ""
            },
            {
                "messages": [
                    "event-9-msg"
                ],
                "note": ""
            }
        ],
        "events": [
            {
                "i": 5,
                "msg": "You're here! I was afraid you weren't going to come for me this time.",
                "msg-n" : "The comment of developer to give hints to translators",
                "trig": 12
            },
            {
                "i": 9,
                "msg": "HELP!!!",
                "trig": 10
            },
            {
                "by-event": [
                    5
                ],
                "i": 12,
                "msg": "Bowser: Say goodbye to your precious princess."
            }
        ],
        "npc": [
            {
                "i": 2,
                "t": 71,
                "talk": "Hahaha. You're not getting in here. This door has been magically sealed. You would have to gather 10 star pieces to open it. Get lost punk!",
                "talk-n": "Another hint of developer for translators",
                "u": true
            }
        ]
    },
    "bowsers_castle/lunadll.txt": {
        "lines": [
            {
                "i": 2,
                "src": "Welcome to super Bowser castle!"
            }
        ],
        "title": "ztarwuff-PavlovGold/lunadll.txt"
    },
    "world.wldx": {
        "credits": "Redigit\n",
        "credits-n": "Just Redigit\n",
        "levels": [
            {
                "f": "star_road.lvlx",
                "i": 0,
                "tit": "Star Road Station"
            },
            {
                "f": "frost_warp.lvlx",
                "i": 1,
                "tit": "Warp to Mushroom Castle"
            },
            {
                "f": "nimbus_land.lvlx",
                "i": 2,
                "tit": "Nimbus Land"
            }
        ],
        "title": "The Princess Cliche"
    }
}
```

## Translations (used by game to apply translated phrases during the level/world loading):
```json
{
    " episode_title": "Клише принцессы",
    " episode_world": "world.wldx",
    "bowsers_castle.lvlx": {
        "events": [
            {
                "i": 5,
                "msg": "О, ты здесь! А то я уже боялась, что ты не придёшь за мной.",
                "orig": "You're here! I was afraid you weren't going to come for me this time."
            },
            {
                "i": 9,
                "msg": "СПАСИТЕ!!!",
                "orig": "HELP!!!"
            },
            {
                "i": 12,
                "msg": "Боузер: Скажи \"прощай\" своей дорогой принцессе.",
                "orig": "Bowser: Say goodbye to your precious princess."
            }
        ],
        "npc": [
            {
                "i": 2,
                "orig": "Hahaha. You're not getting in here. This door has been magically sealed. You would have to gather 10 star pieces to open it. Get lost punk!",
                "talk": "Хахаха. Ты сюда не вой-\nдёшь. Эта дверь запечатана магически. Ты должен соб-\nрать 10 звёзд, чтобы она открылась. Проваливай, сопляк!"
            }
        ]
    },
    "bowsers_castle/lunadll.txt": {
        "lines": [
            {
                "i": 2,
                "src": "Welcome to super Bowser castle!",
                "tr": "Добро пожаловать в супер-замок Боузера!",
                "u": true
            }
        ],
        "title": ""
    },
    "world.wldx": {
        "credits": "Redigit\n\nПеревод на русский:\nWohlstand",
        "credits-n": "Redigit\n\nПеревод на русский:\nWohlstand",
        "credits-orig": "Redigit\n",
        "levels": [
            {
                "i": 0,
                "orig": "Star Road Station",
                "tit": "Станция Звёздной тропы"
            },
            {
                "i": 1,
                "orig": "Warp to Mushroom Castle",
                "tit": "Путь к грибному замку"
            },
            {
                "i": 2,
                "orig": "Nimbus Land",
                "tit": "Земля нимбов"
            }
        ],
        "title": "Клише принцессы",
        "title-orig": "The Princess Cliche"
    }
}
```

**Reply by LoveBodhi** [reply]
This is important information so I need to write a Chinese version of this at later.
